File: src/main/java/com/unihogsoft/scalag/vulkan/core/Instance.java
package com.unihogsoft.scalag.vulkan.core;

import com.unihogsoft.scalag.vulkan.utility.VulkanAssertionError;
import com.unihogsoft.scalag.vulkan.utility.VulkanObject;
import org.lwjgl.BufferUtils;
import org.lwjgl.PointerBuffer;
import org.lwjgl.system.MemoryUtil;
import org.lwjgl.vulkan.VkApplicationInfo;
import org.lwjgl.vulkan.VkInstance;
import org.lwjgl.vulkan.VkInstanceCreateInfo;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import static org.lwjgl.system.MemoryUtil.*;
import static org.lwjgl.system.MemoryUtil.memFree;
import static org.lwjgl.vulkan.EXTDebugReport.VK_EXT_DEBUG_REPORT_EXTENSION_NAME;
import static org.lwjgl.vulkan.VK10.*;
import static org.lwjgl.vulkan.VK10.VK_SUCCESS;

/**
 * @author MarconZet
 * Created 13.04.2020
 */
public class Instance extends VulkanObject {
    private static final String[] VALIDATION_LAYERS_INSTANCE_EXTENSIONS = {VK_EXT_DEBUG_REPORT_EXTENSION_NAME};
    private static final String[] INSTANCE_EXTENSIONS = {};
    private static final String[] VALIDATION_LAYERS = {"VK_LAYER_KHRONOS_validation"};
    private final boolean enableValidationLayers;

    private VkInstance instance;

    public Instance() {
        this.enableValidationLayers = System.getProperty("debug") != null;
        create();
    }

    @Override
    protected void init() {
        VkApplicationInfo appInfo = VkApplicationInfo.calloc()
                .sType(VK_STRUCTURE_TYPE_APPLICATION_INFO)
                .pNext(NULL)
                .pApplicationName(memUTF8("ScalaG MVP"))
                .pEngineName(memUTF8("ScalaG Computing Engine"))
                .applicationVersion(VK_MAKE_VERSION(0, 1, 0))
                .engineVersion(VK_MAKE_VERSION(0, 1, 0))
                .apiVersion(VK_API_VERSION_1_0);

        PointerBuffer ppEnabledExtensionNames = getInstanceExtensions();
        PointerBuffer ppEnabledLayerNames = getValidationLayers();

        VkInstanceCreateInfo pCreateInfo = VkInstanceCreateInfo.calloc()
                .sType(VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO)
                .pNext(NULL)
                .pApplicationInfo(appInfo)
                .ppEnabledExtensionNames(ppEnabledExtensionNames)
                .ppEnabledLayerNames(ppEnabledLayerNames);
        PointerBuffer pInstance = BufferUtils.createPointerBuffer(1);
        int err = vkCreateInstance(pCreateInfo, null, pInstance);
        long instance = pInstance.get(0);
        if (err != VK_SUCCESS) {
            throw new VulkanAssertionError("Failed to create VkInstance", err);
        }
        this.instance = new VkInstance(instance, pCreateInfo);

        memFree(appInfo.pApplicationName());
        memFree(appInfo.pEngineName());
        memFree(ppEnabledExtensionNames);
        memFree(ppEnabledLayerNames);
        appInfo.free();
        pCreateInfo.free();
    }

    private PointerBuffer getValidationLayers() {
        if (enableValidationLayers) {
            PointerBuffer validationLayers = memAllocPointer(VALIDATION_LAYERS.length);
            for (String layer : VALIDATION_LAYERS) {
                validationLayers.put(memASCII(layer));
            }
            return validationLayers.flip();
        } else {
            return null;
        }
    }

    private PointerBuffer getInstanceExtensions() {
        List<String> extensions = Arrays.asList(INSTANCE_EXTENSIONS);
        if (enableValidationLayers)
            extensions.addAll(Arrays.asList(VALIDATION_LAYERS_INSTANCE_EXTENSIONS));

        PointerBuffer ppEnabledExtensionNames = memAllocPointer(extensions.size());
        extensions.forEach(x -> ppEnabledExtensionNames.put(memASCII(x)));
        return ppEnabledExtensionNames.flip();
    }

    @Override
    protected void close() {
        vkDestroyInstance(instance, null);
    }

    public VkInstance get() {
        return instance;
    }
}

File: src/main/scala/com/scalag/vulkan/core/Instance.scala
package com.scalag.vulkan.core

import com.scalag.vulkan.VulkanContext.ValidationLayer
import com.scalag.vulkan.util.Util.{check, pushStack}
import com.scalag.vulkan.util.VulkanObject
import org.lwjgl.system.MemoryStack
import org.lwjgl.system.MemoryUtil.NULL
import org.lwjgl.vulkan.*
import org.lwjgl.vulkan.EXTDebugReport.VK_EXT_DEBUG_REPORT_EXTENSION_NAME
import org.lwjgl.vulkan.KHRPortabilityEnumeration.{VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR, VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME}
import org.lwjgl.vulkan.VK10.*
import org.lwjgl.vulkan.VK13.*

import scala.collection.mutable
import scala.jdk.CollectionConverters.given
import scala.util.chaining.*

/** @author
  *   MarconZet Created 13.04.2020
  */
object Instance {
  val ValidationLayersExtensions: Seq[String] = List(VK_EXT_DEBUG_REPORT_EXTENSION_NAME)
  val MacOsExtensions: Seq[String] = List(VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME)

  lazy val (extensions, layers): (Seq[String], Seq[String]) = pushStack { stack =>
    val ip = stack.ints(1)
    vkEnumerateInstanceLayerProperties(ip, null)
    val availableLayers = VkLayerProperties.malloc(ip.get(0), stack)
    vkEnumerateInstanceLayerProperties(ip, availableLayers)

    vkEnumerateInstanceExtensionProperties(null.asInstanceOf[String], ip, null)
    val instance_extensions = VkExtensionProperties.malloc(ip.get(0), stack)
    vkEnumerateInstanceExtensionProperties(null.asInstanceOf[String], ip, instance_extensions)

    val extensions = instance_extensions.iterator().asScala.map(_.extensionNameString())
    val layers = availableLayers.iterator().asScala.map(_.layerNameString())
    (extensions.toSeq, layers.toSeq)
  }

  lazy val version: Int = VK.getInstanceVersionSupported
}

class Instance(enableValidationLayers: Boolean) extends VulkanObject {

  private val instance: VkInstance = pushStack { stack =>

    val appInfo = VkApplicationInfo
      .calloc(stack)
      .sType$Default()
      .pNext(NULL)
      .pApplicationName(stack.UTF8("ScalaG MVP"))
      .pEngineName(stack.UTF8("ScalaG Computing Engine"))
      .applicationVersion(VK_MAKE_VERSION(0, 1, 0))
      .engineVersion(VK_MAKE_VERSION(0, 1, 0))
      .apiVersion(Instance.version)

    val ppEnabledExtensionNames = getInstanceExtensions(stack)
    val ppEnabledLayerNames = {
      val layers = enabledLayers
      val pointer = stack.callocPointer(layers.length)
      layers.foreach(x => pointer.put(stack.ASCII(x)))
      pointer.flip()
    }

    val pCreateInfo = VkInstanceCreateInfo
      .calloc(stack)
      .sType$Default()
      .flags(VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR)
      .pNext(NULL)
      .pApplicationInfo(appInfo)
      .ppEnabledExtensionNames(ppEnabledExtensionNames)
      .ppEnabledLayerNames(ppEnabledLayerNames)
    val pInstance = stack.mallocPointer(1)
    check(vkCreateInstance(pCreateInfo, null, pInstance), "Failed to create VkInstance")
    new VkInstance(pInstance.get(0), pCreateInfo)
  }

  lazy val enabledLayers: Seq[String] = List
    .empty[String]
    .pipe { x =>
      if (Instance.layers.contains(ValidationLayer) && enableValidationLayers) ValidationLayer +: x
      else if (enableValidationLayers)
        println("Validation layers requested but not available")
        x
      else x
    }

  def get: VkInstance = instance

  override protected def close(): Unit =
    vkDestroyInstance(instance, null)

  private def getInstanceExtensions(stack: MemoryStack) = {
    val extensions = mutable.Buffer.from(Instance.MacOsExtensions)
    if (enableValidationLayers)
      extensions.addAll(Instance.ValidationLayersExtensions)

    val ppEnabledExtensionNames = stack.callocPointer(extensions.size)
    extensions.foreach(x => ppEnabledExtensionNames.put(stack.ASCII(x)))
    ppEnabledExtensionNames.flip()
  }
}
